src/esp32/hcsr04.rs

```diff
@@ -0,0 +1,289 @@
+    // We require a board in order to ensure that ISR setup has completed
+    _board: BoardType,
+
+    // The PinDriver to control the pin that triggers issuing a pulse.
+    trigger_pin: PinDriver<'static, AnyIOPin, Output>,
+
+    // The PinDriver used to listen for digital interrupts and measure
+    // the length of the echo pulse.
+    echo_interrupt_pin: PinDriver<'static, AnyIOPin, Input>,
+
+    // How long we will wait for an echo pulse before concluding that there is no
+    // obstacle in range. Defaults to 50ms.
+    timeout: Duration,
+
+    // The notification channel used to wait on a result being posted from the ISR.
+    interrupt_notification: Notification,
+
+    // State which we share with the ISR.
+    isr_shared_state: Arc<IsrSharedState>,
+}
+
+impl HCSR04Sensor {
+    pub fn from_config(cfg: ConfigType, deps: Vec<Dependency>) -> anyhow::Result<SensorType> {
+        let board = registry::get_board_from_dependencies(deps).ok_or(anyhow::anyhow!(
+            "HCSR04Sensor: failed to get board from dependencies"
+        ))?;
+
+        let trigger_pin = cfg.get_attribute::<i32>("trigger_pin").map_err(|e| {
+            anyhow::anyhow!(
+                "HCSR04Sensor: failed to get `trigger_pin from configuration`: {:?}",
+                e
+            )
+        })?;
+
+        let echo_interrupt_pin = cfg
+            .get_attribute::<i32>("echo_interrupt_pin")
+            .map_err(|e| {
+                anyhow::anyhow!(
+                    "HCSR04Sensor: failed to get `echo_interrupt_pin` from configuration: {:?}",
+                    e
+                )
+            })?;
+
+        let timeout = cfg.get_attribute::<u32>("timeout_ms").map_or_else(
+            |e| match e {
+                AttributeError::KeyNotFound(_) => Ok(None),
+                _ => Err(anyhow::anyhow!(
+                    "HCSR04Sensor: error handling `timeout_ms` value: {:?}",
+                    e
+                )),
+            },
+            |v| Ok(Some(Duration::from_millis(v as u64))),
+        )?;
+
+        Ok(Arc::new(Mutex::new(HCSR04Sensor::new(
+            board,
+            trigger_pin,
+            echo_interrupt_pin,
+            timeout,
+        )?)))
+    }
+
+    fn new(
+        board: BoardType,
+        trigger_pin: i32,
+        echo_interrupt_pin: i32,
+        timeout: Option<Duration>,
+    ) -> anyhow::Result<HCSR04Sensor> {
+        // TODO: Move this into micro-rdk and unify with pin.rs
+        init_isr_alloc_flags(esp_idf_hal::interrupt::InterruptType::Iram.into());
+        enable_isr_service()?;
+
+        let notification = Notification::new();
+        let notifier = notification.notifier();
+
+        let mut sensor = Self {
+            _board: board,
+            trigger_pin: PinDriver::output(unsafe { AnyIOPin::new(trigger_pin) })?,
+            echo_interrupt_pin: PinDriver::input(unsafe { AnyIOPin::new(echo_interrupt_pin) })?,
+            timeout: timeout.unwrap_or(Duration::from_millis(50)),
```

Comment:
The RDK defaults `timeout_ms` to 1000ms which is far too long. Also, we are blocking in `wait` for up to this duration, so it really doesn't make sense. The range of the sensor is 4m so 50ms is more than enough. I think I might consider clipping any configured value higher than, say, 100ms.

Reply:
This is now clamped between 100us and 100ms since those are below and above the documented min and max echo response pulse times.